### ISSUE TITLE
Shielded transfer over any secure channel without z-address

### ISSUE THREAD

**Issue by tromer** [zcash/zcash#2432]
Enable sending shielded ZEC over any unidirectional secure channel, without knowing the recipient's z-address.

Motivating scenarios:
* send ZEC to someone via Open-PGP encrypted email, knowing nothing more than the recipient's email address and OpenPGP key
* send ZEC to someone over any channel encrypted via PKI
* send ZEC to someone over Signal or WhatsApp

Out-of-band note distribution (#801) over the secure channel gets us mostly there, by letting the recipient know v,r,rho,memo. The one missing component is that the recipients needs to know a_sk matching the a_pk used in the JoinSplit.

A simple approach is for the sender to create a fresh a_sk and send this, too, in the secure channel. This means the sender can cancel the transfer and withdraw the ZEC back into his own wallet, up until the funds are withdrawn by the receiver (and adequately confirmed). But this seems OK for some scenarios, like opportunistic donations or transfer between trusting parties. And it may actually be a desired feature because it avoids fund loss if the channel is unreliable (e.g., email eaten by spam filter, or Signal device key is lost).

Need to analyze interaction with selective disclosure.

**Comment by daira** [comment]
Note that if the sender knows a_sk then they can compute the nullifier. This means that any transaction in which the recipient spends the note will be identifiable by the sender, which may not be desired if the recipient was otherwise anonymous to the sender. (Whether that is the case depends on how the channel was established.)

**Comment by str4d** [comment]
In the opportunistic donation scenario that would be fine, because the appearance of the nullifier would signal to the sender's software that the window in which the funds could be reclaimed has closed. The UX around this would need to be carefully constructed however, to ensure that recipients are aware that they should not be using the address corresponding to the `a_sk` anywhere else (ie. it should be ephemeral at the UX level).

**Comment by str4d** [comment]
There's also a much simpler way to implement this, in the ephemeral `a_sk` case: generate a new SpendingKey, send funds to its PaymentAddress, then send the SpendingKey to the recipient over the unidirectional secure channel (instead of the encrypted contents). We would not want to have this look like an actual SpendingKey though, for the above UX reason. The reason I mention this is because this specific feature is tangential to out-of-band notes - in the latter, the user does not want their note data permanently stored in the block chain (and trusts the secure OOB channel more), whereas here, the user wants to send without knowing the recipient's z-address (but having the secure OOB channel).

**Comment by daira** [comment]
Right, in that case you probably don't care about anonymity of the recipient to the sender. But more generally it is a weakening of anonymity properties, which would have to be explained to users (and we already have a lot of subtle things to explain).

**Comment by str4d** [comment]
> Right, in that case you probably don't care about anonymity of the recipient to the sender. But more generally it is a weakening of anonymity properties, which would have to be explained to users (and we already have a lot of subtle things to explain).

Correct. It is a potentially useful trade-off however, if the secure OOB channel is in fact unidirectional (with the nullifier visibility being the useful confirmation in the reverse direction). For any bidirectional secure OOB channel, it's more secure to just have the sender ask the recipient to create a z-address, so the motivations for this issue centre on whether unidirectional channels are a large-enough use case. All three scenarios @tromer gave technically use a bidirectional channel, though there's no guarantee the recipient will ever respond.

One possibly-tangential effect of this proposal would be to lower the perceived barrier to getting started with Zcash. For a bidirectional case, the recipient has to:

- Read the incoming message asking for a z-address
- Install a wallet
- Generate a z-address
- Send the z-address to the sender
- Wait for them to send the ZEC (and either be notified by the sender, or the wallet itself)

For the unidirectional case, the recipient has to:

- Read the incoming message containing the "transfer string"
- Install a wallet (assume all wallets support this feature)
- Paste in the "transfer string"

**Comment by radix42** [comment]
oh wow @str4d I can't say how much I love that use case

**Comment by nathan-at-least** [comment]
I also love the unidirectional UX in [@str4d's comment](https://github.com/zcash/zcash/issues/2432#issuecomment-306961131).
